Blinding/src/TrackDigiExtractor_module.cc

```diff
@@ -0,0 +1,122 @@
+// Ed Callaghan
+// Extract digis which participate in a prefit track
+// September 2024
+
+// stl
+#include <string>
+
+// art
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Principal/Event.h"
+
+// cetlib_except
+#include "cetlib_except/exception.h"
+
+// fhiclcpp
+#include "fhiclcpp/types/Atom.h"
+#include "fhiclcpp/types/Comment.h"
+#include "fhiclcpp/types/Name.h"
+
+// mu2e
+#include "Offline/RecoDataProducts/inc/KalSeed.hh"
+#include "Offline/RecoDataProducts/inc/StrawHit.hh"
+#include "Offline/RecoDataProducts/inc/StrawDigi.hh"
+#include "Offline/RecoDataProducts/inc/TrkStrawHitSeed.hh"
+
+namespace mu2e{
+  class TrackDigiExtractor: public art::EDProducer{
+    public:
+      struct Config{
+        // track
+        fhicl::Atom<art::InputTag> kalseed_tag{
+          fhicl::Name("KalSeedCollection"),
+          fhicl::Comment("art::InputTag of KalSeeds to extract digis from")
+        };
+
+        // tracker
+        fhicl::Atom<art::InputTag> strawhit_tag{
+          fhicl::Name("StrawHitCollection"),
+          fhicl::Comment("art::InputTag of StrawHits associated with KalSeed")
+        };
+        fhicl::Atom<art::InputTag> strawdigi_tag{
+          fhicl::Name("StrawDigiCollection"),
+          fhicl::Comment("art::InputTag of StrawDigis associated with StrawHits")
+        };
+
+        // calorimeter...
+      };
+
+      using Parameters = art::EDProducer::Table<Config>;
+      TrackDigiExtractor(const Parameters&);
+
+    protected:
+      art::InputTag _kalseed_tag;
+      art::InputTag _strawhit_tag;
+      art::InputTag _strawdigi_tag;
+
+    private:
+      void produce(art::Event&);
+  };
+
+  TrackDigiExtractor::TrackDigiExtractor(const Parameters& config):
+      art::EDProducer(config),
+      _kalseed_tag(config().kalseed_tag()),
+      _strawhit_tag(config().strawhit_tag()),
+      _strawdigi_tag(config().strawdigi_tag()){
+    this->consumes<KalSeedCollection>(_kalseed_tag);
+    this->consumes<StrawHitCollection>(_strawhit_tag);
+    this->consumes<StrawDigiCollection>(_strawdigi_tag);
+    this->consumes<StrawDigiADCWaveformCollection>(_strawdigi_tag);
+    this->produces<StrawDigiCollection>();
+    this->produces<StrawDigiADCWaveformCollection>();
+  }
+
+  void TrackDigiExtractor::produce(art::Event& event){
+    // data-product lookups
+    auto kalseeds = event.getValidHandle<KalSeedCollection>(_kalseed_tag);
+    auto strawhits = event.getValidHandle<StrawHitCollection>(_strawhit_tag);
+    auto strawdigis = event.getValidHandle<StrawDigiCollection>(_strawdigi_tag);
+    auto strawadcss = event.getValidHandle<StrawDigiADCWaveformCollection>(_strawdigi_tag);
+
+    // sanity checks
+    if (strawhits->size() != strawdigis->size()){
```

Comment:
This test relies on this module's config being consistent with what was used when fitting the tracks.
 
A way to ensure consistency is to use the KKTrack instead of KalSeed; KKTrack can have Ptrs back to underlying collections, but it cannot be persisted, due to payload that includes std::Ptrs. We can discuss if that makes sense in this context.

Reply:
Hmmm, ok. The current implementation does depend on this check passing, as e.g. `StrawHit`s and `StrawDigi`s do not contain any ptrs to each other. To be sure that both for a given hit are included, we have to assume that the collections in which they are contained are aligned (and the best check we have for that is same-length).

Reply:
There is a long-term issue that we need to improve the robustness of the digi->hit->trkhit indexing using ProductPtr, which didn't exist when the original codes were written. That's beyond the scope of this PR. We can deal with this code when we address that issue in general.

Reply:
Sounds good to me.